src/engine/interop/specification.md

```diff
@@ -0,0 +1,130 @@
+- `parentHash`: `DATA`, 32 Bytes
+- `coinbase`:  `DATA`, 20 Bytes
+- `stateRoot`: `DATA`, 32 Bytes
+- `receiptRoot`: `DATA`, 32 Bytes
+- `logsBloom`: `DATA`, 256 Bytes
+- `random`: `DATA`, 32 Bytes
+- `blockNumber`: `QUANTITY`
+- `gasLimit`: `QUANTITY`
+- `gasUsed`: `QUANTITY`
+- `timestamp`: `QUANTITY`
+- `extraData`: `DATA`
+- `baseFeePerGas`: `QUANTITY`
+- `blockHash`: `DATA`, 32 Bytes
+- `transactions`: `Array of DATA` - Array of transaction objects, each object is a byte list (`DATA`) representing `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)
+
+## Core
+
+### engine_preparePayload
+
+#### Parameters
+1. `Object` - The payload attributes:
+- `parentHash`: `DATA`, 32 Bytes - hash of the parent block
+- `timestamp`: `QUANTITY` - value for the `timestamp` field of the new payload
+- `random`: `DATA`, 32 Bytes - value for the `random` field of the new payload
+- `feeRecipient`: `DATA`, 20 Bytes - suggested value for the `coinbase` field of the new payload
+
+#### Returns
+1. `payloadId`: `QUANTITY` - identifier of the payload building process
+
+#### Specification
+
+1. Given provided field value set client software **SHOULD** build the initial version of the payload which has an empty transaction set.
+
+2. Client software **SHOULD** start the process of updating the payload. The strategy of this process is implementation dependent. The default strategy would be to keep the transaction set up-to-date with the state of local mempool.
+
+3. Client software **SHOULD** stop the updating process either by finishing to serve the [**`engine_getPayload`**](#engine_getPayload) call with the same `payloadId` value or when [`SECONDS_PER_SLOT`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#time-parameters-1) (currently set to 12 in the Mainnet configuration) seconds have passed since the point in time identified by the `timestamp` parameter.
+
+4. Client software **MUST** set the payload field values according to the set of parameters passed in the call to this method with exception for the `feeRecipient`. The `coinbase` field value **MAY** stem from what is specified by the `feeRecipient` parameter.
+
+### engine_getPayload
+
+#### Parameters
+1. `payloadId`: `QUANTITY` - Identifier of the payload building process
+
+#### Returns
+`Object|Error` - Either instance of [`ExecutionPayload`](#ExecutionPayload) or an error
+
+#### Specification
+
+1. Given the `payloadId` client software **MUST** respond with the most recent version of the payload that is available in the corresponding building process at the time of receiving the call.
+
+2. The call **MUST** be responded with error code `2` (`Action not allowed`) if a building process identified by the `payloadId` doesn't exist.
```

Comment:
This isn't an obvious response.  Could we add a new error code `4` (`Unknown ID`) to make this error explicit?

Reply:
I was not sure what is the rule of thumb of adding new error codes and just picked the one that already exists with the following description:
> Should be used when some action is not allowed, e.g. preventing an action, while another depending action is processing on, like sending again when a confirmation popup is shown to the user

Do you know if each custom error code supposed to be used only in one place or semantically similar places of the client software?

Reply:
It's not so much "not allowed" as "not possible" (due to the ID being unknown), so although the best of the current options it isn't a great answer.

I don't know what the process is to add to error codes, but I would certainly try to do so rather than overload an existing error code.

Reply:
I've added `4: Missing resource` and used it here and in a couple of other places.